Shift to `workflow-session`

CBroz1 opened the pull request:
Added:
- Subset of example data for animal/lab/session - no examples for `genotyping` yet
- Expanded existing notebook
- Future second notebook pending nwb export function PR
- Updated install instructions per refactor point to `datajoint-elements`
- Integration tests for ingestion and pipeline structure
- integration test for `element-animal` part tables
- verbosity option for integration test
- docker-compose-test files

Future PRs:
- Integration tests for nwb export
- Re-add export notebook
- Genotyping data

CBroz1 committed:
Fixed link formatting, adding parens

[title]link.com -> [title](link.com)

CBroz1 committed:
Fixed link formatting, adding parens

[title]link.com -> [title](link.com)

CBroz1 committed:
Merge branch 'datajoint:main' into patch-1

CBroz1 committed:
Merge PR, minor alt branch formatting edits

Merge branches of my fork for minor text edits

CBroz1 committed:
initial commit migrate wf-anim->wf-sess

CBroz1 committed:
partial revision, fake data, ingest functs

CBroz1 committed:
Adding NB for export example

CBroz1 committed:
Adding NB for export example

CBroz1 committed:
more detail in NB output

CBroz1 committed:
Update to reflect sess not depend on lab

CBroz1 committed:
update readme/diagrams, add activate(genotyping)

CBroz1 committed:
fixing readme typos

CBroz1 commented:
Major changes ready for review. Minor fixes to export notebook pending commits to `element-session`

CBroz1 committed:
start integration test add, WIP

CBroz1 committed:
coverage tests: ingest & pipeline

CBroz1 commented on .gitignore:
```diff
@@ -1,18 +1,9 @@
-# vscode
```
Reorganized the gitignore to be more compact

CBroz1 commented on CHANGELOG.md:
```diff
@@ -0,0 +1,22 @@
++ Added notebooks
+
+[0.0.0b1]: https://github.com/datajoint/workflow-animal/tree/f6d6a3353aae83ca13ff9fcc536017eb34c18f90
+[0.0.0a1]: https://github.com/datajoint/workflow-animal/tree/0d3d3c970056ff4c243d17cf4f738f48268d80ad
```
- Do workflows get changelogs? I would retroactively consider the 3/18 and 3/24 changes as releases. 
- Considered present changes substantial enough to shift to 0.0.1

kabilar replied:
We have previously implemented changelogs, tags, and releases for the `elements`, but not the `workflows`.  I would be in favor of having changelogs and tags for the `workflows`.  In which case, I would suggest keeping the versions of the `element` and corresponding `workflow` independent from each other.

Tagging for further discussion @ttngu207 @dimitri-yatsenko @iamamutt.

CBroz1 committed:
remove pynwb version fix

CBroz1 committed:
Remove notes file

CBroz1 commented:
@kabilar - I added integration tests for the primary ingest and pipeline functions here. Ready to review pending an additional nwb export integration test. Do you have privileges to invite Ben to the conversation? Is such a test within his scope of work?

dimitri-yatsenko commented on tests/__init__.py:
```diff
@@ -0,0 +1,238 @@
+
+import os
+import pytest
+import pandas as pd
```
I think the use of pandas here is unjustified. All the operations where it's used can be just more cleanly  accomplished with python datatypes: dicts and lists.

CBroz1 replied:
I removed pandas and numpy with most recent commit

dimitri-yatsenko commented on tests/__init__.py:
```diff
@@ -0,0 +1,238 @@
+@pytest.fixture
+def lab_proj_csv():
+    """ Create a 'projects.csv' file"""
+    input_lab_proj = pd.DataFrame(columns=['project','project_description','repositoryurl',
```
writing a sample CSV file with sample data can be made much simpler. Consider simply writing a multiline string into a file.

CBroz1 replied:
Sure! I was following the model presented in [workflow-array-ephys/tests/__init__.py](https://github.com/datajoint/workflow-array-ephys/blob/main/tests/__init__.py), but easy enough to remove pandas here. Should the same edits on the other workflows? I'm not seeing anywhere else within workflow-array-ephys/tests that pandas is used.

dimitri-yatsenko commented on tests/__init__.py:
```diff
@@ -0,0 +1,238 @@
+    lab_projusers_csv_path.unlink()                       # delete csv file after use
+
+@pytest.fixture
+def lab_pubs_csv():
```
For PEP8, use complete English words for function names.

CBroz1 replied:
Thanks for this note! I'll make this adjustment systematically

dimitri-yatsenko commented on tests/__init__.py:
```diff
@@ -0,0 +1,238 @@
+    lab_keyw_csv_path.unlink()                  # delete csv file after use
+
+@pytest.fixture
+def lab_prot_csv():
```
PEP 8: use English words for identifiers.

kabilar commented:
Thanks @CBroz1.  I appreciate your work on this.  If its not too difficult, could you use git to rename files (`git mv <old name> <new name>`).  It will make it easier to track changes.

kabilar commented:
@CBroz1, how about we add the integration tests for the NWB export function with a subsequent pull request?

CBroz1 committed:
reverting folder name before gitmv; for git comparisons

CBroz1 committed:
folder rename w/gitmv; for git comparisons

CBroz1 committed:
rm numpy/pandas; PEP8 full eng words

dimitri-yatsenko commented on README.md:
```diff
@@ -9,7 +9,7 @@ This repository provides demonstrations for:
 
 ## Workflow architecture
-The lab and animal management workflow presented here uses components from two DataJoint elements (element-lab, element-animal and element-session) assembled together into a functional workflow.
+The lab and subject management workflow presented here uses components from two DataJoint elements (element-lab, element-animal and element-session) assembled together into a functional workflow.
```
```suggestion
The lab and experiment subject management workflow presented here uses components from two DataJoint elements (element-lab, element-animal and element-session) assembled together into a functional workflow.
```

CBroz1 committed:
PEP8 E501,F401; Diagram aspect ration & new tables

PEP8: Added IDE linter to fix E501 (line max to 79) and
      F401 (unused imports).
Diagrams: Updated aspect ratio for legibility, preserved old
					See datajoint-python/issues/975
					Added new element-lab tables. See element-lab/pull/11

CBroz1 committed:
Update README subject->experiment subject

CBroz1 committed:
minor: remove redundant import

CBroz1 committed:
Merge branch, resolve conflict

CBroz1 committed:
merging branch conflicts

CBroz1 committed:
Merge branch 'main' of https://github.com/CBroz1/workflow-session

CBroz1 committed:
Minor; resolve conflicts, same notebooks

CBroz1 committed:
remove old diagrams, update README links

CBroz1 committed:
Removed lab.Project refs to slices, viruses, etc. - see details

As discussed in [element-session issue #9](https://github.com/datajoint/element-session/issues/9),
 moving pharmacology, slices, surgery, and virus to session level. Here, removed from
- projects.csv
- ingest.py
- notebooks

2_Explore_Export.ipynb will need more changes after [element-session PR #7](https://github.com/datajoint/element-session/pull/7) is finalized

CBroz1 committed:
Removed lab.Project refs to slices, etc. from integration test

CBroz1 committed:
Version. Update CHANGELOG. notebook: remove specifics

- Version: move from __init__.py to version.py per updated convention
- Version: bump
- CHANGELOG: update with rename, integration tests
- Notebook: remove my username from output. Needs future updates after [element-session PR](https://github.com/datajoint/element-session/pull/7)

CBroz1 commented:
Missing: While I've added a docker-compose, I'm not familiar with using docker to run the relevant integration tests, and have run the current tests though terminal pytest. @kabilar - Do you have time to demonstrate the docker process you've used for other elements, or should we reserve that step for when we're further along with github actions?

kabilar commented:
Hi @CBroz1, definitely.  During our meeting tomorrow, let's review how to run the integration tests locally with Docker.  Thanks.

CBroz1 committed:
Docker integration test successful. See details.

Changes:
README: Refactor to point to `datajoint-elements` for installation
Docker-compose*: follow new `workflow-array-ephys` model
notebook/2_Explore_Export: remove pending separate nwb PR
tests/__init__.py: add mkdir for csv subdirectories
tests/test_export.py: remove pending nwb PR
workflow_session/pipeline.py: comment out genotyping schema info

CBroz1 committed:
Remove diagrams, link to elements in readme

kabilar commented on CHANGELOG.md:
```diff
@@ -0,0 +1,25 @@
++ Integration tests
+
+### Changed
++ Name `element-animal` -> `element-session`
```
```suggestion
+ Name `workflow-animal` -> `workflow-session`
```

kabilar commented on CHANGELOG.md:
```diff
@@ -0,0 +1,25 @@
+### Added
++ Added notebooks
+
+[0.0.0b1]: https://github.com/datajoint/workflow-animal/tree/f6d6a3353aae83ca13ff9fcc536017eb34c18f90
+[0.0.0a1]: https://github.com/datajoint/workflow-animal/tree/0d3d3c970056ff4c243d17cf4f738f48268d80ad
```
```suggestion
[0.0.0b1]: https://github.com/datajoint/workflow-session/tree/f6d6a3353aae83ca13ff9fcc536017eb34c18f90
[0.0.0a1]: https://github.com/datajoint/workflow-session/tree/0d3d3c970056ff4c243d17cf4f738f48268d80ad
```

kabilar commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,3 @@
+# Contribution Guidelines
+
+This project follows the [DataJoint Contribution Guidelines](https://docs.datajoint.io/python/community/02-Contribute.html). Please reference the link for more full details.
```
```suggestion
This project follows the [DataJoint Contribution Guidelines](https://docs.datajoint.org/python/community/02-Contribute.html). Please reference the link for more full details.
```

kabilar commented on setup.py:
```diff
@@ -6,26 +6,27 @@
 Build a workflow for lab management and animal metadata using DataJoint Elements
-+ [elements-lab](https://github.com/datajoint/elements-lab)
-+ [elements-animal](https://github.com/datajoint/elements-animal)
++ [elements-lab](https://github.com/datajoint/element-lab)
++ [elements-animal](https://github.com/datajoint/element-animal)
++ [elements-session](https://github.com/datajoint/element-session)
```
```suggestion
+ [element-lab](https://github.com/datajoint/element-lab)
+ [element-animal](https://github.com/datajoint/element-animal)
+ [element-session](https://github.com/datajoint/element-session)
```

kabilar commented on setup.py:
```diff
@@ -6,26 +6,27 @@
     long_description=long_description,
-    author='DataJoint NEURO',
+    author='DataJoint',
     author_email='info@vathes.com',
```
```suggestion
    author_email='info@datajoint.com',
```

kabilar commented on requirements.txt:
```diff
@@ -1,4 +1,6 @@
 datajoint>=0.13.0
 element-lab
 element-animal
+element-session
 ipykernel
+pynwb
```
Let's lock these versions.

CBroz1 committed:
Apply suggestions from @kabilar's code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
pin requirements

CBroz1 committed:
Move diagrams to element. Revise table names per elem-lab change

CBroz1 committed:
add integration tests for Subject part tables

CBroz1 committed:
Fix coverage output, Add verbosity option. See details

docker-compose-test.yaml: fix -cov hyphen->underscore
requirements.txt: version bump element-lab to new version with ProjectKeywords table
setup.py: restructure following wf-ephys format
tests/__init__.py: add verbosity boolean to clean integration test output
workflow_session/ingest.py: add optional verbosity function arguments
workflow_session/paths.py: remove unused root_dir function

regarding verbosity: I found removing print statements made it easier to see which tests passed

kabilar commented on LICENSE:
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2020 DataJoint NEURO
+Copyright (c) 2021 DataJoint
```
```suggestion
Copyright (c) 2022 DataJoint
```

kabilar commented on apt-requirements.txt:
```diff
@@ -0,0 +1 @@
+locales-all
```
```suggestion
git
locales-all
```

kabilar replied:
Also, please rename `apt-requirements.txt` to `apt_requirements.txt`.

kabilar commented on CHANGELOG.md:
```diff
@@ -0,0 +1,24 @@
++ First beta release
+
+
+## 0.0.0a1 - 2021-03-18
```
```suggestion
## [0.0.0a1] - 2021-03-18
```

kabilar commented on CHANGELOG.md:
```diff
@@ -0,0 +1,24 @@
++ Name `workflow-animal` -> `workflow-session`
+
+
+## 0.0.0b1 - 2021-03-24
```
```suggestion
## [0.0.0b1] - 2021-03-24
```

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
rename apt_requirements.txt

CBroz1 commented:
Thanks for your review @kabilar. I applied the suggestions and changed the name of `apt_requirements.txt`

kabilar commented:
Hi @CBroz1, it looks like `apt-requirements.txt` was moved to `apt_requirements.txt/apt-requirements.txt`.

CBroz1 committed:
rename apt_requirements

CBroz1 committed:
rename typo

CBroz1 commented:
> Hi @CBroz1, it looks like `apt-requirements.txt` was moved to `apt_requirements.txt/apt-requirements.txt`.

Oops! good catch. fixed

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,35 @@
+WORKDIR /main/workflow-session
+
+# Option 1 - Install DataJoint's remote fork of the workflow and elements
+# RUN git clone https://github.com/datajoint/workflow-session.git /main/workflow-session
```
```suggestion
# RUN git clone https://github.com/datajoint/workflow-session.git /main/
```

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,35 @@
+# RUN pip install git+https://github.com/cbroz1/element-lab.git
+# RUN pip install git+https://github.com/<user>/element-animal.git
+# RUN pip install git+https://github.com/<user>/element-session.git
+# RUN git clone https://github.com/<user>/workflow-session.git /main/workflow-session
```
```suggestion
# RUN git clone https://github.com/<user>/workflow-session.git /main/
```

kabilar commented on docker-compose-test.yaml:
```diff
@@ -0,0 +1,45 @@
+    env_file: .env
+    image: workflow_session:0.0.0b2
+    environment:
+      - DJ_HOST=workflow-session_db_1
```
```suggestion
      - DJ_HOST=db
```

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

kabilar commented on workflow_session/pipeline.py:
```diff
@@ -0,0 +1,34 @@
+Experimenter = lab.User
+session.activate(db_prefix + 'session', linking_module=__name__)
+
+# genotyping.activate(db_prefix + 'genotyping', linking_module=__name__)
```
```suggestion
genotyping.activate(db_prefix + 'genotyping', db_prefix + 'subject', linking_module=__name__)
```

kabilar commented on workflow_session/pipeline.py:
```diff
@@ -0,0 +1,34 @@
+from element_animal.subject import Subject
+# from element_animal.genotyping import Sequence, BreedingPair, Cage,\
+#                                       SubjectCaging, GenotypeTest
+from element_lab.lab import Source, Lab, Protocol, User, Project, ProjectUser
+from element_lab.lab import ProjectKeywords, ProjectPublication, ProjectSourceCode
```
```suggestion
from element_lab.lab import Source, Lab, Protocol, User, Project, ProjectUser, \
                                                  ProjectKeywords, ProjectPublication, ProjectSourceCode
```

kabilar commented on workflow_session/pipeline.py:
```diff
@@ -0,0 +1,34 @@
+#                                       SubjectCaging, GenotypeTest
+from element_lab.lab import Source, Lab, Protocol, User, Project, ProjectUser
+from element_lab.lab import ProjectKeywords, ProjectPublication, ProjectSourceCode
+from element_session.session import Session, SessionDirectory, SessionExperimenter
+from element_session.session import SessionNote, ProjectSession
```
```suggestion
from element_session.session import Session, SessionDirectory, SessionExperimenter, \
                                                                 SessionNote, ProjectSession
```

kabilar commented on workflow_session/pipeline.py:
```diff
@@ -0,0 +1,34 @@
+from element_session import session
+
+from element_animal.subject import Subject
+# from element_animal.genotyping import Sequence, BreedingPair, Cage,\
+#                                       SubjectCaging, GenotypeTest
```
```suggestion
from element_animal.genotyping import Sequence, BreedingPair, Cage,\
                                      SubjectCaging, GenotypeTest
```

kabilar commented on workflow_session/pipeline.py:
```diff
@@ -0,0 +1,34 @@
+           'ProjectUser', 'Session', 'SessionDirectory', 'SessionExperimenter',
+           'SessionNote', 'ProjectSession']
+
+# Activate "lab", "subject", "session" schema -------------
```
```suggestion
# Activate "lab", "subject", "session", "genotyping" schemas -------------
```

CBroz1 replied:
My thought process was that we shouldn't try to demonstrate use of the genotyping schema without example data. Should we activate and keep empty even without the didactic tools to help the user explore?

CBroz1 committed:
Apply suggestions from code review: revise import formatting

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
Apply suggestions from code review: activate genotyping shema

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,35 @@
+FROM datajoint/djlab:py3.8-debian
+
+USER root
+RUN apt-get update -y
+RUN apt-get install git -y
+
+USER anaconda
```
```suggestion
USER anaconda:anaconda
```

kabilar replied:
Git will now be installed through the `apt_requirements.txt` (see [DataJoint image notes](https://datajoint.github.io/miniconda3-docker/#features)).

kabilar commented on README.md:
```diff
@@ -1,177 +1,47 @@
 
 ## Workflow architecture
-The lab and animal management workflow presented here uses components from two DataJoint elements, element-lab, and element-animal assembled together to a functional workflow.
+The lab and experiment subject management workflow presented here uses components from two DataJoint elements (element-lab, element-animal and element-session) assembled together into a functional workflow.
```
```suggestion
The lab and experiment subject management workflow presented here uses components from three DataJoint elements (element-lab, element-animal and element-session) assembled together into a functional workflow.
```

kabilar commented on Dockerfile.dev:
```diff
@@ -0,0 +1,29 @@
+FROM datajoint/djlab:py3.8-debian
+
+USER root
+RUN apt-get update -y
+RUN apt-get install git -y
+
+USER anaconda
```
```suggestion
USER anaconda:anaconda
```

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,35 @@
+
+# Option 2 - Install user's remote fork of element and workflow
+#            or an unreleased version of the element
+# RUN pip install git+https://github.com/cbroz1/element-lab.git
```
```suggestion
# RUN pip install git+https://github.com/<user>/element-lab.git
```

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,35 @@
+# RUN rm -f /main/workflow-session/dj_local_conf.json
+
+# Install the workflow
+RUN pip install /main/workflow-session
```
```suggestion
RUN pip install -e /main/workflow-session
```

kabilar commented on user_data/lab/sources.csv:
```diff
@@ -0,0 +1,2 @@
+source, source_name, contact_details, source_description
+Provider1, Example Provider, +44 1632 960663 / Example@Provider.com, UK-based supplier of lab subjcts mus musculus
```
```suggestion
Provider1, Example Provider, +44 1632 960663 / Example@Provider.com, UK-based supplier of lab subjects mus musculus
```

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 commented:
Thanks for your review @kabilar - I've committed all your suggested changes

kabilar commented on docker-compose-test.yaml:
```diff
@@ -0,0 +1,45 @@
+      context: ../
+      dockerfile: ./workflow-session/Dockerfile.test
+    env_file: .env
+    image: workflow_session:0.0.0b2
```
```suggestion
    image: workflow_session_test:0.0.0b2
```

kabilar commented on Dockerfile.test:
```diff
@@ -0,0 +1,31 @@
+FROM datajoint/djlab:py3.8-debian
+
+USER anaconda:anaconda
```
```suggestion
USER anaconda:anaconda

COPY ./workflow-session/apt_requirements.txt /tmp/
RUN /entrypoint.sh echo "Installed dependencies."
```

kabilar commented on Dockerfile.dev:
```diff
@@ -0,0 +1,25 @@
+FROM datajoint/djlab:py3.8-debian
+
+USER anaconda:anaconda
```
```suggestion
USER anaconda:anaconda

COPY ./workflow-session/apt_requirements.txt /tmp/
RUN /entrypoint.sh echo "Installed dependencies."
```

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,112 @@
+
+    ingest_general(csvs, tables, skip_duplicates=skip_duplicates, verbose=verbose)
+
+    # TODO: add allele and genotyping data
```
```suggestion
```

kabilar replied:
Added to #12.

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,112 @@
+        :param tables: list of datajoint tables with ()
+        :param verbose: print number inserted (i.e., table length change)
+    """
+    for insert, table in zip(csvs, tables):
+        with open(insert, newline='') as f:
```
```suggestion
    for csv_filepath, table in zip(csvs, tables):
        with open(csv_filepath, newline='') as f:
```

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,112 @@
+            print(f'\n---- Inserting {insert_len} entry(s) '
+                  + f'into {table.table_name} ----')
+
+    # Future enhancement: permit embedded lists
+    # Currently requires a csv to be listed multiple times
+    # if feeding multiple tables. Could instead take lists of lists.
```
```suggestion
```

kabilar replied:
I am inclined to make these GitHub Issues, but we can keep here if you think it is more appropriate.

CBroz1 replied:
Sure thing - added issue #17

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,112 @@
+        if verbose:
+            prev_len = len(table)
+        table.insert(data, skip_duplicates=skip_duplicates,
+                     # Ignore Extra because some CSVs feed/mult tables
```
```suggestion
                     # Ignore extra fields because some CSVs feed multiple tables
```

CBroz1 committed:
Apply all suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

kabilar commented on tests/test_pipeline_generation.py:
```diff
@@ -0,0 +1,43 @@
+
+    # test connection Subject -> schema children
+    session_tbl, _, subject_line_tbl, subject_protocol_tbl, subject_source_tbl, \
+        subject_strain_tbl, subject_user_tbl, subject_cull_table, subject_death_tbl,\
```
```suggestion
        subject_strain_tbl, subject_user_tbl, subject_cull_tbl, subject_death_tbl,\
```

CBroz1 replied:
Are you suggesting here that we should spell out each instance of `tbl`->`table`? I adopted this format from [wf-ephys](https://github.com/datajoint/workflow-array-ephys/blob/main/tests/test_pipeline_generation.py), but non-words is not PEP8 compliant. I can make the shift systematically

kabilar replied:
Thanks @CBroz1.  No need to make that PEP8 changes in this pull request.  I was actually just trying to make all variables with the same suffix (`tbl`).

kabilar commented on tests/test_pipeline_generation.py:
```diff
@@ -0,0 +1,43 @@
+    assert subject_source_tbl.full_table_name == subject.Subject.Source.full_table_name
+    assert subject_strain_tbl.full_table_name == subject.Subject.Strain.full_table_name
+    assert subject_user_tbl.full_table_name == subject.Subject.User.full_table_name
+    assert subject_cull_table.full_table_name == \
```
```suggestion
    assert subject_cull_tbl.full_table_name == \
```

kabilar commented on tests/__init__.py:
```diff
@@ -0,0 +1,335 @@
+
+
+"""
+# FUTURE FIXTURES, pending sharable example data
+lab_sources_content=["source, source_name, contact_details,\
+                      source_description"]
+lab_sources_csv_path = pathlib.Path('./tests/user_data/lab/sources.csv')
+write_csv(lab_sources_content,lab_sources_csv_path)
+yeild lab_sources_content,lab_sources_csv_path
+lab_sources_csv_path.unlink()
+subject_allele_content=["allele, allele_standard_name,zygosity"]
+subject_allele_csv_path = pathlib.Path('./tests/user_data/subject/allele.csv')
+write_csv(subject_allele_content,subject_allele_csv_path)
+yeild subject_allele_content,subject_allele_csv_path
+subject_allele_csv_path.unlink()
+subject_genotyping_content=[]
+subject_genotyping_csv_path = pathlib.Path('./tests/user_data/subject/\
+                                            genotyping.csv')
+write_csv(subject_genotyping_content,subject_genotyping_csv_path)
+yeild subject_genotyping_content,subject_genotyping_csv_path
+subject_genotyping_csv_path.unlink()
+subject_line_content=["line, line_description, target_phenotype, is_active"]
+subject_line_csv_path = pathlib.Path('./tests/user_data/subject/line.csv')
+write_csv(subject_line_content,subject_line_csv_path)
+yeild subject_line_content,subject_line_csv_path
+subject_line_csv_path.unlink()
+subject_source_content=["allele,source_identifier,source_url,\
+                         expression_data_url"]
+subject_source_csv_path = pathlib.Path('./tests/user_data/subject/source.csv')
+write_csv(subject_source_content,subject_source_csv_path)
+yeild subject_source_content,subject_source_csv_path
+subject_source_csv_path.unlink()
+subject_strain_content=["strain, strain_standard_name, strain_desc"]
+subject_strain_csv_path = pathlib.Path('./tests/user_data/subject/strain.csv')
+write_csv(subject_strain_content,subject_strain_csv_path)
+yeild subject_strain_content,subject_strain_csv_path
+subject_strain_csv_path.unlink()
+subject_subjects_csv_path = pathlib.Path('./tests/user_data/subject/\
+                                          subjects.csv')
+"""
```
```suggestion
```

kabilar replied:
Let's move this to a GitHub Issue.

kabilar commented on README.md:
```diff
@@ -1,177 +1,47 @@
 
-![lab](images/lab_diagram.svg)
+![element-lab](
+https://github.com/datajoint/element-lab/raw/main/images/element_lab_diagram.svg)
```
```suggestion
https://github.com/datajoint/element-lab/raw/main/images/lab_diagram.svg)
```

kabilar commented on README.md:
```diff
@@ -1,177 +1,47 @@
-
-## Development mode installation
++ The installation instructions can be found at [datajoint-elements/install.md](
+     https://github.com/datajoint/datajoint-elements/blob/main/install.md).
```
```suggestion
     https://github.com/datajoint/datajoint-elements/blob/main/gh-pages/docs/install.md).
```

kabilar commented on README.md:
```diff
@@ -1,177 +1,47 @@
-Please refer to the [DataJoint Labbook page](https://github.com/datajoint/datajoint-labbook) for instructions to set it up.
-
-## Development mode installation
++ The installation instructions can be found at [datajoint-elements/install.md](
```
```suggestion
+ The installation instructions can be found in the [datajoint-elements repository](
```

kabilar commented on README.md:
````diff
@@ -1,177 +1,47 @@
-    ```
++ Please refer to the following workflow-specific
+[Jupyter notebooks](/notebooks) for an in-depth explanation of how to run the
+workflow ([1-Explore_Workflow.ipynb](notebooks/1_Explore_Workflow.ipynb).
````
```suggestion
workflow ([1-Explore_Workflow.ipynb](notebooks/1_Explore_Workflow.ipynb)).
```

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,108 @@
+    By default, uses data from workflow_session/user_data/lab/
+    :param lab_csv_path:      relative path of lab csv
+    :param project_csv_path:  relative path of project csv
+    :param pubs_csv_path:     relative path of publication csv
+    :param keyw_csv_path:     relative path of keyword csv
```
```suggestion
    :param publication_csv_path:     relative path of publication csv
    :param keyword_csv_path:     relative path of keyword csv
```

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,108 @@
+    :param pubs_csv_path:     relative path of publication csv
+    :param keyw_csv_path:     relative path of keyword csv
+    :param protocol_csv_path: relative path of protocol csv
+    :param users_csv_path:    relative path of users csv
```
```suggestion
    :param users_csv_path:    relative path of users csv
    :param project_user_csv_path: relative path of project users csv
```

kabilar commented on workflow_session/ingest.py:
```diff
@@ -0,0 +1,108 @@
+    """
+    Inserts data from a sessions csv into corresponding session schema tables
+    By default, uses data from workflow_session/user_data/session/
+    :param session_csv_path:     relative path of subject csv
```
```suggestion
    :param session_csv_path:     relative path of session csv
```

kabilar commented on tests/__init__.py:
```diff
@@ -0,0 +1,335 @@
+pathlib.Path('./tests/user_data/session').mkdir(exist_ok=True)
+pathlib.Path('./tests/user_data/subject').mkdir(exist_ok=True)
+
+# ------------------ GENERAL FUCNTION ------------------
```
```suggestion
# ------------------ GENERAL FUNCTIONS ------------------
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -0,0 +1,95 @@
+'''Tests ingestion into schema tables: Lab, Subject, Session
+    1. Assert length of populating data from __init__
+    2. Assert exact matches of inserted data fore key tables
```
```suggestion
    2. Assert exact matches of inserted data for key tables
```

kabilar commented on tests/test_ingest.py:
```diff
@@ -0,0 +1,95 @@
+        assert (lab.Protocol & {'protocol': protocol_values[0]}
+                ).fetch1('protocol_type') == protocol_values[1]
+
+    # Does not have example data:
+    # assert len(lab.Source()) == 0
```
```suggestion
```

kabilar replied:
Let's add to a GitHub Issue.

kabilar commented on tests/test_ingest.py:
```diff
@@ -0,0 +1,95 @@
+        assert (subject.Subject.User & {'subject': subject_values[0]}
+                ).fetch1('user') == subject_values[2]
+
+    # Does not have example data:
+    # assert len(genotyping.Sequence()) == 0
```
```suggestion
```

kabilar replied:
Let's add to a GitHub Issue.

kabilar commented on notebooks/1_Explore_Workflow.ipynb:
```diff
@@ -0,0 +1,1542 @@
+   "id": "c47691a0-b016-4092-a5ad-fefff93c54dd",
+   "metadata": {},
+   "source": [
+    "For more documentation of insert, please refer to [DataJoint Docs](https://docs.datajoint.io/python/manipulation/1-Insert.html) and [DataJoint playground](https://playground.datajoint.io/)"
```
```suggestion
    "For more documentation of insert, please refer to [DataJoint Docs](https://docs.datajoint.org/python/manipulation/1-Insert.html) and [DataJoint CodeBook](https://codebook.datajoint.io/)"
```

kabilar commented on notebooks/1_Explore_Workflow.ipynb:
```diff
@@ -0,0 +1,1542 @@
+   "id": "c510fe4d-09ed-472f-830f-4401bd6830d0",
+   "metadata": {},
+   "source": [
+    "(Workflow needs continued development to import geotyping tables)"
```
```suggestion
```

kabilar commented on notebooks/1_Explore_Workflow.ipynb:
```diff
@@ -0,0 +1,1542 @@
+   "id": "13f8a8ed-2656-46d8-82ba-cdf130c4873e",
+   "metadata": {},
+   "source": [
+    "## Insert into Manual and Lookup tables with Graphical User Interface"
```
```suggestion
    "## Insert into Manual and Lookup tables with a Graphical User Interface"
```

kabilar commented on notebooks/1_Explore_Workflow.ipynb:
```diff
@@ -0,0 +1,1542 @@
+   "id": "4775dd80-8a54-47b7-a9ba-99995db9ff1a",
+   "metadata": {},
+   "source": [
+    "DataJoint also provides a Graphical User Interface [DataJoint Labbook](https://github.com/datajoint/datajoint-labbook) to support manual data insertions into DataJoint workflows. ![DataJoint Labbook preview](https://github.com/datajoint/datajoint-labbook/blob/master/docs/sphinx/_static/images/walkthroughDemoOptimized.gif)"
```
```suggestion
    "DataJoint also provides a graphical user interface ([DataJoint LabBook](https://github.com/datajoint/datajoint-labbook)) to support manual data insertions into DataJoint workflows. ![DataJoint LabBook preview](https://github.com/datajoint/datajoint-labbook/blob/master/docs/sphinx/_static/images/walkthroughDemoOptimized.gif)"
```

kabilar reviewed (changes requested):
Thanks @CBroz1!  I appreciate you taking the time to put together the example data, ingestion functions, pytests, and other updates.  A few minor suggestions above.

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
Apply suggestions from code review

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

CBroz1 committed:
Update notebooks/1_Explore_Workflow.ipynb

Co-authored-by: Kabilar Gunalan <kabilar@datajoint.com>

kabilar reviewed (approved):
Thanks @CBroz1!
